examples/memcached.md

```diff
@@ -0,0 +1,197 @@
+`SIGUSR1`. The state file describes the content of the memory file and
+is used during a restart by memcached to decide whether it can use the
+existing cached data.
+
+The [example
+deployment](../deploy/kustomize/memcached/persistent/memcached-persistent.yaml)
+uses a stateful set because that can automatically create persistent
+volumes for each instance. The shell wrapper around memcached
+translates the normal [`SIGTERM` shutdown
+signal](https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods)
+into `SIGUSR1`.
+
+Deploying like that becomes less flexible because the memcached pods
+can no longer move freely between nodes. Instead, each pod has to be
+restarted on the node where its volume was created. There are also
+[several caveats for memcached in this
+mode](https://github.com/memcached/memcached/wiki/WarmRestart#caveats)
+that admins and application developers must be aware of.
+
+This example can also be kustomized. By default it uses the
+[`pmem-csi-sc-ext4` storage
+class](../deploy/common/pmem-storageclass-ext4.yaml), which must be
+installed in the cluster. Here we just use the defaults, in particular
+the default namespace:
+
+```
+$ kubectl apply -f deploy/kustomize/memcached/persistent/memcached-persistent.yaml
+service/pmem-memcached created
+statefulset.apps/pmem-memcached created
+```
+
+Beware that at the moment, the volumes need to be removed manually
+after removing the stateful set. A [request to automate
+that](https://github.com/kubernetes/kubernetes/issues/55045) has been
+pending for a while.
+
+```
+$ kubectl delete -f deploy/kustomize/memcached/persistent/memcached-persistent.yaml
+service "pmem-memcached" deleted
+statefulset.apps "pmem-memcached" deleted
+$ kubectl delete pvc -l app.kubernetes.io/name=pmem-memcached
+persistentvolumeclaim "memcached-data-volume-pmem-memcached-0" deleted
+```
+
+We can verify that memcached really does a warm restart by storing some
+data, removing the instance and then starting it again.
+```
+$ kubectl apply -f deploy/kustomize/memcached/persistent/memcached-persistent.yaml
+service/pmem-memcached created
+statefulset.apps/pmem-memcached created
+$ kubectl wait --for=condition=Ready pods -l app.kubernetes.io/name=pmem-memcached
+pod/pmem-memcached-0 condition met
+$ pod=$(kubectl get pods -l app.kubernetes.io/name=pmem-memcached -o jsonpath={..metadata.name})
+$ kubectl port-forward $pod 11211 & kubectl=$!
+Forwarding from 127.0.0.1:11211 -> 11211
+Forwarding from [::1]:11211 -> 11211
+$ telnet 127.0.0.1 11211
+Trying 127.0.0.1...
+Connected to 127.0.0.1.
+Escape character is '^]'.
+Handling connection for 11211
+set demo_key 0 500000 10
+I am PMEM.
+STORED
+quit
+Connection closed by foreign host.
+$ kubectl scale --replicas=0 statefulset/pmem-memcached
+statefulset.apps/pmem-memcached scaled
+$ kubectl wait --for delete pod/$pod
+Error from server (NotFound): pods "pmem-memcached-0" not found
```

Comment:
Is there a race condition to worry about here?  My pods did not take that long to terminate so I don't understand the need for the wait commands.

Reply:
Yes, there's a race without it. The stateful set might get scaled down and back up before the controller reacts to the change, in which case the original pod keeps running.